File: cli/test/kb/conflicts_test.exs
defmodule Kb.ConflictsTest do
  use ExUnit.Case

  alias Kb.Conflicts

  @moduletag :tmp_dir

  describe "lifecycle/1" do
    test "picks up YAML frontmatter" do
      draft = """
      ---
      lifecycle: draft
      ---
      # Page
      """

      reviewed = """
      ---
      lifecycle: reviewed
      ---
      # Page
      """

      verified = """
      ---
      lifecycle: verified
      ---
      # Page
      """

      assert Conflicts.lifecycle(draft) == :draft
      assert Conflicts.lifecycle(reviewed) == :reviewed
      assert Conflicts.lifecycle(verified) == :verified
    end

    test "returns :none when no frontmatter" do
      assert Conflicts.lifecycle("# Page\n\nBody") == :none
    end
  end

  describe "scan_page?/1" do
    test "rejects drafts and accepts everything else", %{tmp_dir: tmp} do
      draft = Path.join(tmp, "d.md")
      reviewed = Path.join(tmp, "r.md")
      plain = Path.join(tmp, "p.md")

      File.write!(draft, "---\nlifecycle: draft\n---\n# x")
      File.write!(reviewed, "---\nlifecycle: reviewed\n---\n# x")
      File.write!(plain, "# x")

      refute Conflicts.scan_page?(draft)
      assert Conflicts.scan_page?(reviewed)
      assert Conflicts.scan_page?(plain)
    end
  end

  describe "extract_claims/1" do
    test "parses a Key Claims section into claim structs", %{tmp_dir: tmp} do
      path = Path.join(tmp, "src.md")

      File.write!(path, """
      # Source
      ## Key Claims
      - JWT tokens expire after 24 hours — c-1ee0207b
      - Refresh tokens are one-time use — c-fcba1172

      ## Other
      unrelated
      """)

      claims = Conflicts.extract_claims(path)
      assert length(claims) == 2
      assert Enum.at(claims, 0).chunk_id == "c-1ee0207b"
      assert Enum.at(claims, 0).source == "src"
      assert Enum.at(claims, 0).text =~ "JWT"
    end
  end

  describe "run/1 with heuristic provider" do
    test "planted contradiction across sources lands in report", %{tmp_dir: tmp} do
      kb_dir = setup_kb_with_contradiction(tmp)

      {:ok, %{findings: findings, report_path: path}} =
        Conflicts.run(kb_dir: kb_dir, provider: "heuristic", date: ~D[2026-04-23])

      assert length(findings) >= 1
      assert File.exists?(path)

      body = File.read!(path)
      assert body =~ "Contradictions — 2026-04-23"
      assert body =~ "concept: attention.md"
      assert body =~ "c-aaaaaaaa"
      assert body =~ "c-bbbbbbbb"
      assert body =~ "Claim A"
      assert body =~ "Claim B"
    end

    test "agreeing sources produce no false positive", %{tmp_dir: tmp} do
      kb_dir = setup_kb_with_agreement(tmp)

      {:ok, %{findings: findings, report_path: path}} =
        Conflicts.run(kb_dir: kb_dir, provider: "heuristic", date: ~D[2026-04-23])

      assert findings == []
      assert File.exists?(path)
      body = File.read!(path)
      assert body =~ "No contradictions"
    end

    test "skips draft concept pages", %{tmp_dir: tmp} do
      kb_dir = setup_kb_with_contradiction(tmp)

      # Flip concept to draft
      concept = Path.join([kb_dir, "wiki", "concepts", "attention.md"])
      original = File.read!(concept)
      File.write!(concept, "---\nlifecycle: draft\n---\n" <> original)

      {:ok, %{findings: findings}} =
        Conflicts.run(kb_dir: kb_dir, provider: "heuristic", date: ~D[2026-04-23])

      assert findings == []
    end
  end

  defp setup_kb_with_contradiction(tmp) do
    kb_dir = Path.join(tmp, "kb")
    concepts = Path.join([kb_dir, "wiki", "concepts"])
    sources = Path.join([kb_dir, "wiki", "sources"])
    File.mkdir_p!(concepts)
    File.mkdir_p!(sources)

    File.write!(Path.join(concepts, "attention.md"), """
    # Attention

    > Related sources

    ## Details
    See [vaswani](../sources/vaswani.md) and [flash](../sources/flash.md).

    ## Provenance
    - [vaswani.md](../sources/vaswani.md) — c-aaaaaaaa
    - [flash.md](../sources/flash.md) — c-bbbbbbbb
    """)

    File.write!(Path.join(sources, "vaswani.md"), """
    # Vaswani 2017

    ## Key Claims
    - Attention complexity is quadratic with dense patterns — c-aaaaaaaa
    """)

    File.write!(Path.join(sources, "flash.md"), """
    # FlashAttention

    ## Key Claims
    - Attention complexity is linear with sparse patterns — c-bbbbbbbb
    """)

    kb_dir
  end

  defp setup_kb_with_agreement(tmp) do
    kb_dir = Path.join(tmp, "kb")
    concepts = Path.join([kb_dir, "wiki", "concepts"])
    sources = Path.join([kb_dir, "wiki", "sources"])
    File.mkdir_p!(concepts)
    File.mkdir_p!(sources)

    File.write!(Path.join(concepts, "jwt.md"), """
    # JWT

    ## Details
    See [a](../sources/a.md) and [b](../sources/b.md).
    """)

    File.write!(Path.join(sources, "a.md"), """
    # A

    ## Key Claims
    - JWT tokens expire after 24 hours — c-11111111
    """)

    File.write!(Path.join(sources, "b.md"), """
    # B

    ## Key Claims
    - JWT tokens are valid for 24 hours — c-22222222
    """)

    kb_dir
  end
end

File: packages/inferrers/test/conflicts.test.ts
import { afterEach, beforeEach, describe, expect, test } from 'bun:test';
import { existsSync, mkdirSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import {
  crossSourcePairs,
  detectConflicts,
  extractClaims,
  findReferencedSources,
  heuristicComparator,
  lifecycle,
  llmComparator,
  renderReport,
  shouldScan,
  type ClaimComparator,
  type Finding,
} from '../src/conflicts.ts';
import type { LLMProvider } from '@kb/llm';

let tmp: string;

beforeEach(() => {
  tmp = mkdtempSync(join(tmpdir(), 'kb-conflicts-'));
  mkdirSync(join(tmp, 'kb/wiki/concepts'), { recursive: true });
  mkdirSync(join(tmp, 'kb/wiki/sources'), { recursive: true });
});

afterEach(() => {
  rmSync(tmp, { recursive: true, force: true });
});

function srcPage(name: string, claims: { text: string; id: string }[]): string {
  const path = join(tmp, 'kb/wiki/sources', `${name}.md`);
  const body = `# ${name}\n\n## Key Claims\n\n${claims
    .map((c) => `- ${c.text} — c-${c.id}`)
    .join('\n')}\n`;
  writeFileSync(path, body);
  return path;
}

function conceptPage(name: string, links: string[]): string {
  const path = join(tmp, 'kb/wiki/concepts', `${name}.md`);
  const body =
    `# ${name}\n\n` +
    links.map((l) => `- See [${l}](../sources/${l}.md)`).join('\n') +
    '\n';
  writeFileSync(path, body);
  return path;
}

describe('lifecycle / scan gating', () => {
  test('frontmatter lifecycle wins over body text', () => {
    expect(lifecycle('---\nlifecycle: draft\n---\nbody\n')).toBe('draft');
    expect(lifecycle('---\nlifecycle: reviewed\n---\nbody\n')).toBe('reviewed');
  });

  test('blockquote lifecycle outside frontmatter is detected', () => {
    expect(lifecycle('> lifecycle: verified\nhello')).toBe('verified');
  });

  test('no lifecycle => "none"', () => {
    expect(lifecycle('# no fm here')).toBe('none');
  });

  test('shouldScan skips draft, accepts none/reviewed/verified', () => {
    const draft = join(tmp, 'd.md');
    const review = join(tmp, 'r.md');
    writeFileSync(draft, '---\nlifecycle: draft\n---\n');
    writeFileSync(review, '---\nlifecycle: reviewed\n---\n');
    expect(shouldScan(draft)).toBe(false);
    expect(shouldScan(review)).toBe(true);
  });
});

describe('claim extraction', () => {
  test('parses key claim bullets with chunk ids', () => {
    const p = srcPage('alpha', [
      { text: 'attention is quadratic', id: 'a1b2c3d4' },
      { text: 'cache hit rates matter', id: 'deadbeef' },
    ]);
    const claims = extractClaims(p);
    expect(claims).toHaveLength(2);
    expect(claims[0]!.chunkId).toBe('c-a1b2c3d4');
    expect(claims[0]!.text).toBe('attention is quadratic');
    expect(claims[0]!.source).toBe('alpha');
  });

  test('crossSourcePairs ignores same-source pairs', () => {
    const claims = [
      { text: 'a', source: 'X', sourcePath: 'x', chunkId: 'c-1' },
      { text: 'b', source: 'X', sourcePath: 'x', chunkId: 'c-2' },
      { text: 'c', source: 'Y', sourcePath: 'y', chunkId: 'c-3' },
    ];
    const pairs = crossSourcePairs(claims);
    expect(pairs).toHaveLength(2);
    for (const [a, b] of pairs) expect(a.source).not.toBe(b.source);
  });
});

describe('findReferencedSources', () => {
  test('matches relative sources/<slug>.md links', () => {
    srcPage('alpha', []);
    srcPage('beta', []);
    const c = conceptPage('xforms', ['alpha', 'beta', 'missing']);
    const refs = findReferencedSources(c, tmp);
    const names = refs.map((p) => p.split('/').pop()).sort();
    expect(names).toEqual(['alpha.md', 'beta.md']);
  });
});

describe('heuristic comparator', () => {
  test('flags antonym pairs sharing topic', async () => {
    const a = { text: 'attention patterns are dense across tokens', source: 's1', sourcePath: '', chunkId: 'c-1' };
    const b = { text: 'attention patterns are sparse across tokens', source: 's2', sourcePath: '', chunkId: 'c-2' };
    expect(await heuristicComparator.compare(a, b)).toBe('disagree');
  });

  test('returns agree for aligned claims', async () => {
    const a = { text: 'attention is quadratic in tokens', source: 's1', sourcePath: '', chunkId: 'c-1' };
    const b = { text: 'attention scales quadratically with tokens', source: 's2', sourcePath: '', chunkId: 'c-2' };
    expect(await heuristicComparator.compare(a, b)).toBe('agree');
  });

  test('returns unknown when topic overlap is too small', async () => {
    const a = { text: 'unicorns gallop in the meadow', source: 's1', sourcePath: '', chunkId: 'c-1' };
    const b = { text: 'kernels schedule across cores', source: 's2', sourcePath: '', chunkId: 'c-2' };
    expect(await heuristicComparator.compare(a, b)).toBe('unknown');
  });

  test('flags negation asymmetry on shared topic', async () => {
    const a = { text: 'tokens are cached across requests', source: 's1', sourcePath: '', chunkId: 'c-1' };
    const b = { text: 'tokens are not cached across requests', source: 's2', sourcePath: '', chunkId: 'c-2' };
    expect(await heuristicComparator.compare(a, b)).toBe('disagree');
  });
});

describe('detectConflicts pipeline', () => {
  test('finds planted contradiction and writes report', async () => {
    const a = srcPage('paperA', [
      { text: 'attention patterns are dense across tokens', id: 'aaaaaaaa' },
    ]);
    const b = srcPage('paperB', [
      { text: 'attention patterns are sparse across tokens', id: 'bbbbbbbb' },
    ]);
    conceptPage('attention', ['paperA', 'paperB']);
    void a;
    void b;

    const result = await detectConflicts({ kbRoot: tmp, date: new Date('2026-05-21') });
    expect(result.findings).toHaveLength(1);
    expect(result.findings[0]!.concept).toBe('attention');
    expect(result.reportPath).toContain('contradictions-2026-05-21.md');
    const body = readFileSync(result.reportPath!, 'utf8');
    expect(body).toContain('## concept: attention.md');
    expect(body).toContain('c-aaaaaaaa');
    expect(body).toContain('c-bbbbbbbb');
  });

  test('emits "no contradictions" report when none found', async () => {
    srcPage('one', [{ text: 'cats purr softly', id: '11111111' }]);
    srcPage('two', [{ text: 'cats purr softly', id: '22222222' }]);
    conceptPage('cats', ['one', 'two']);
    const result = await detectConflicts({ kbRoot: tmp, date: new Date('2026-05-21') });
    expect(result.findings).toHaveLength(0);
    expect(readFileSync(result.reportPath!, 'utf8')).toContain('No contradictions found');
  });

  test('skips draft source pages', async () => {
    const a = srcPage('paperA', [
      { text: 'attention patterns are dense across tokens', id: 'aaaaaaaa' },
    ]);
    const b = srcPage('paperB', [
      { text: 'attention patterns are sparse across tokens', id: 'bbbbbbbb' },
    ]);
    // mark paperB draft
    writeFileSync(b, '---\nlifecycle: draft\n---\n' + readFileSync(b, 'utf8'));
    conceptPage('attention', ['paperA', 'paperB']);
    void a;
    const result = await detectConflicts({ kbRoot: tmp, date: new Date('2026-05-21') });
    expect(result.findings).toHaveLength(0);
  });

  test('skips when no concepts dir', async () => {
    rmSync(join(tmp, 'kb/wiki/concepts'), { recursive: true });
    const result = await detectConflicts({ kbRoot: tmp });
    expect(result.findings).toHaveLength(0);
    expect(result.reportPath).toBeNull();
  });

  test('llmComparator uses provider verdict', async () => {
    const provider: LLMProvider = {
      name: 'fake',
      async complete() {
        return 'DISAGREE';
      },
    };
    const cmp: ClaimComparator = llmComparator(provider);
    srcPage('paperA', [{ text: 'X is true', id: 'aaaaaaaa' }]);
    srcPage('paperB', [{ text: 'Y is false', id: 'bbbbbbbb' }]);
    conceptPage('thing', ['paperA', 'paperB']);
    const result = await detectConflicts({
      kbRoot: tmp,
      comparator: cmp,
      date: new Date('2026-05-21'),
    });
    expect(result.findings).toHaveLength(1);
  });
});

describe('renderReport', () => {
  test('groups findings by concept', () => {
    const findings: Finding[] = [
      {
        concept: 'b',
        claimA: { text: 't1', source: 'x', sourcePath: '', chunkId: 'c-1' },
        claimB: { text: 't2', source: 'y', sourcePath: '', chunkId: 'c-2' },
      },
      {
        concept: 'a',
        claimA: { text: 'u1', source: 'x', sourcePath: '', chunkId: 'c-3' },
        claimB: { text: 'u2', source: 'y', sourcePath: '', chunkId: 'c-4' },
      },
    ];
    const out = renderReport(findings, new Date('2026-05-21'));
    expect(out.indexOf('## concept: a.md')).toBeLessThan(out.indexOf('## concept: b.md'));
  });
});

// keep used import
void existsSync;
